### PR TITLE
[READY] New clown traitor item: The clown car

### PR DESCRIPTION
**Pull request by ghost**
:cl: Floyd / Qustinnus (Credits to KMC for the sprites)
add: The Clown Car, your 18TC clown-only solution to asshole co-workers
add: Regular car implementation (makes it easier to add more cars if someone actually feels like adding those)
/:cl:

Credits to Goon ofcourse, they had the OG idea.

I've been wanting to add this for a while, and now I have. The clown car is a vehicle you can use if you're a traitor clown to run people over, shove them into your compact trunk, and take them for a ride. If you drive into any walls however, you and everyone you kidnapped, fall out of the car, leaving you somewhat vulnerable.

It also comes with a lube-defense mechanism(tm) which has a 1/3 chance to drop some lube if someone decides to shoot at your sweet ride.

You can also honk your horn or drop all of the drivers if you would like to.

video: https://streamable.com/9b5ww

also i created this while being slightly tired so theres probably still some oversights, as always

TODO:
- [x] Add something to vehicle/sealed that allows you to pick up things you drop inside of them without kicking out the occupants.
- [x] more funny gimmicks
- [x] balance (im bad at this so i might need help with this :^))
- [x] test (>implying)

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
Leave it to the twink to PR the ability to be trapped in a trunk helplessly by an odd man.

**Comment by Qustinnus** [comment]
dont forget the lube

**Comment by MrDoomBringer** [comment]
ahahah holy fuck

**Comment by WJohn** [comment]
Definitely needs a proper clown car honk and not a re-use of the honk mech's honk, but other than that this is amazing.

**Comment by Qustinnus** [comment]
The mech actually re-uses the airhorn sound, I figured the airhorn was usually our "louder bikehorn"

**Comment by Mickyan** [comment]
Shouldn't eject the driver when it crashes imho: too expensive, too funny and relatively innocuous to be stopped so easily and if you kidnapped someone chances are they're going to jack your ride as soon as you get up, I'd say damage the car and temporarily stun the driver instead

Make it so when people are ejected they get thrown out in random directions instead of all dumped on the same tile

**Comment by Qustinnus** [comment]
only clowns can drive the vehicle

**Comment by Mickyan** [comment]
Ah fair enough that makes sense, though if I'm reading this right you can just drive into it with a secway or a scooter to crash it, is that right?

File: code/modules/vehicles/car.dm [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		new /obj/effect/particle_effect/foam(loc) //YEET
+
+/obj/vehicle/sealed/car/clowncar/Bump(atom/movable/M)
+	..()
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		src.visible_message("<span class='danger'>[src] rams into [C] and knocks them down!</span>")
+		C.Knockdown(50)
+		playsound(src, pick('sound/vehicles/clowncar_ram1.ogg', 'sound/vehicles/clowncar_ram2.ogg', 'sound/vehicles/clowncar_ram3.ogg'), 75)
+	else if(!istype(M, /obj/machinery/door) && (istype(M, /obj) || istype(M, /turf/closed)))
```
[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vehicles/car.dm]: why is this not a skateboard subtype

[the remark this answers is not in the document]
Comment (optimumtact) [reply code/modules/vehicles/car.dm]: he might not want all the functionality there.

[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: @optimumtact good point

**Comment by Qustinnus** [comment]
I was considering that, I might make it only crash into things such as walls and structures

**Comment by CosmicScientist** [comment]
Time to drive everyone into the supermatter shard!

**Comment by fluffe9911** [comment]
I would do that getting past doors would be annoying otherwise

**Comment by Qustinnus** [comment]
it already ignores doors

File: code/modules/vehicles/car.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vehicles/car.dm]: that's why it's a subtype not an overwrite :^)

**Comment by fluffe9911** [comment]
Oh noice in that case wooo clown car!

File: code/modules/vehicles/car.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: @ExcessiveUseOfCobblestone vehicles split off at the start depending on if you buckle onto them or are inside of them. youre inside the clown car you buckle to the skateboard

[the remark this answers is not in the document]
Comment (optimumtact) [reply code/modules/vehicles/car.dm]: stupid cobby

**Comment by Kmc2000** [comment]
SPEEDMERGE NOW

**Comment by WJohn** [comment]
carhorn.ogg is better, but it's not *honky* enough. needs a proper clown car honk. something like this https://www.youtube.com/watch?v=x3SxEOvAOEg

File: code/modules/vehicles/car.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vehicles/car.dm]: Then at least modularize the crashing mechanic rather than duplicate reeeeeeeeeeeeeeeee

**Comment by Iamgoofball** [comment]
Just make it use the inserted bike horn's sound and make the car horn a subtype.

**Comment by MrDoomBringer** [comment]
@Qustinnus def make it only crash into walls or glass. Otherwise people will drag around shit (or use secways) for the car to bump into, which makes it pretty much unusable

**Comment by Qustinnus** [comment]
@Iamgoofball wtf good idea

**Comment by GuyonBroadway** [comment]
Legit going to steal this for hippie if it gets denied

**Comment by Deepfreeze2k** [comment]
Only if Clown Ops also get access to this.

**Comment by pizzatiger** [comment]
If you aren't going to keep the part of the goonstation clowncar where the clown car doesn't crash against windows which alot of clown car users use to break/turn  then I suggest you lower the penalty for crashing so it isn't so extreme/deadly

**Comment by Qustinnus** [comment]
alright I'm back from vacay so I'll get to finishing this up

**Comment by Ziiro** [comment]
Couple ideas:
* Bowling pin crashing noise when it crashes and people are forced out of the car.
* Make the horn a lesser version of the HONKBLASTER That knocks people down(super short)/back/their shoes or hats off (god please don't keep the honkblaster's deafening)

**Comment by MrDoomBringer** [comment]
def supporting the bowling pin crashing noise @Ziiro

**Comment by SpaceManiac** [comment]
Stale, ping any maintainer for reopen.

**Comment by Qustinnus** [comment]
@SpaceManiac I'm working on it right now :p currently testing a new commit

**Comment by Qustinnus** [comment]
and was just asking a testmerge too

**Comment by Qustinnus** [comment]
New features:

Automatic sucking of people you run over (Buff)
Inserted horn is used
People fly out in a random direction.

**Comment by Qustinnus** [comment]
you now drop items below the car because fuck adding an inventory or whatever, also new icon for removing key

**Comment by Cw3040** [comment]
this is fine

File: code/modules/vehicles/car.dm [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	..()
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		C.Knockdown(50)
+		C.visible_message("<span class='warning'>[src] rams into [C] and sucks him up!</span>") //fuck off shezza this isn't ERP.
+		mob_forced_enter(C)
+
+		playsound(src, pick('sound/vehicles/clowncar_ram1.ogg', 'sound/vehicles/clowncar_ram2.ogg', 'sound/vehicles/clowncar_ram3.ogg'), 75)
+	else if(istype(M, /turf/closed))
+		src.visible_message("<span class='warning'>[src] rams into [M] and crashes!</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: no src.

File: code/modules/vehicles/vehicle_actions.dm [hunk 3]
```diff
@@ -103,10 +103,53 @@
...
+//CLOWN CAR ACTION DATUMS
+/datum/action/vehicle/sealed/horn
+	name = "Honk Horn"
+	desc = "Honk your classy horn."
+	button_icon_state = "car_horn"
+	var/hornsound = 'sound/items/carhorn.ogg'
+	var/last_honk_time
+
+/datum/action/vehicle/sealed/horn/Trigger()
+	if(world.time - last_honk_time > 20)
```
Comment (kevinz000) [inline code/modules/vehicles/vehicle_actions.dm]: imo this should have the cooldown be a variable and have these two procs not be so duplicated but hey whatever

File: code/modules/vehicles/entered.dm [hunk 4]
```diff
@@ -35,22 +35,74 @@
...
+	to_chat(user, "<span class='notice'>You remove \the [inserted_key] from \the [src].</span>")
+	inserted_key.forceMove(drop_location())
+	user.put_in_hands(inserted_key)
+	inserted_key = null
+	return TRUE
+
+/obj/vehicle/sealed/deconstruct(disassembled = TRUE)
+	. = ..()
+	DumpMobs()
+	explosion(src.loc, 0, 1, 2, 3, 0)
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: src. is not needed and why are you making every vehicle explode by default?

Comment (Qustinnus) [reply code/modules/vehicles/entered.dm]: because its extremely funny

File: code/modules/vehicles/entered.dm [hunk 5]
```diff
@@ -35,22 +35,74 @@
...
+		return
+		to_chat(user, "<span class='notice'>There is no key in \the [src]!</span>")
+	if(!is_occupant(user) || !(occupants[user] & VEHICLE_CONTROL_DRIVE))
+		to_chat(user, "<span class='notice'>You must be driving \the [src] to remove [src]'s key!</span>")
+		return
+	to_chat(user, "<span class='notice'>You remove \the [inserted_key] from \the [src].</span>")
+	inserted_key.forceMove(drop_location())
+	user.put_in_hands(inserted_key)
+	inserted_key = null
+	return TRUE
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: if you're going to return true here the other paths should return false

File: code/modules/vehicles/entered.dm [hunk 6]
```diff
@@ -35,22 +35,74 @@
...
+
+/obj/vehicle/sealed/deconstruct(disassembled = TRUE)
+	. = ..()
+	DumpMobs()
+	explosion(src.loc, 0, 1, 2, 3, 0)
+
+/obj/vehicle/sealed/proc/DumpMobs(randomstep = TRUE)
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: random "step" (throw) shouldn't be default for non clown cars I feel

File: code/modules/vehicles/entered.dm [hunk 7]
```diff
@@ -35,22 +35,74 @@
...
+	for(var/i in occupants)
+		if(iscarbon(i))
+			var/mob/living/carbon/Carbon = i
+			mob_exit(Carbon, null, randomstep)
+			Carbon.Knockdown(40)
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: ?

File: code/modules/vehicles/entered.dm [hunk 8]
```diff
@@ -35,22 +35,74 @@
...
+	return ..()
+
+/obj/vehicle/sealed/proc/remove_key(mob/user)
+	if(!inserted_key)
+		return
+		to_chat(user, "<span class='notice'>There is no key in \the [src]!</span>")
+	if(!is_occupant(user) || !(occupants[user] & VEHICLE_CONTROL_DRIVE))
+		to_chat(user, "<span class='notice'>You must be driving \the [src] to remove [src]'s key!</span>")
+		return
+	to_chat(user, "<span class='notice'>You remove \the [inserted_key] from \the [src].</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: 90% sure \the isn't needed

File: code/modules/vehicles/entered.dm [hunk 9]
```diff
@@ -35,22 +35,74 @@
...
+			to_chat(user, "<span class='notice'>[I] seems to be stuck to your hand!</span>")
+		return
+	return ..()
+
+/obj/vehicle/sealed/proc/remove_key(mob/user)
+	if(!inserted_key)
+		return
+		to_chat(user, "<span class='notice'>There is no key in \the [src]!</span>")
+	if(!is_occupant(user) || !(occupants[user] & VEHICLE_CONTROL_DRIVE))
+		to_chat(user, "<span class='notice'>You must be driving \the [src] to remove [src]'s key!</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: 90% sure \the isn't needed

File: code/modules/vehicles/entered.dm [hunk 10]
```diff
@@ -35,22 +35,74 @@
...
+			inserted_key = I
+		else
+			to_chat(user, "<span class='notice'>[I] seems to be stuck to your hand!</span>")
+		return
+	return ..()
+
+/obj/vehicle/sealed/proc/remove_key(mob/user)
+	if(!inserted_key)
+		return
+		to_chat(user, "<span class='notice'>There is no key in \the [src]!</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: 90% sure \the isn't needed

File: code/modules/vehicles/entered.dm [hunk 11]
```diff
@@ -35,22 +35,74 @@
...
+				inserted_key.forceMove(drop_location())
+			inserted_key = I
+		else
+			to_chat(user, "<span class='notice'>[I] seems to be stuck to your hand!</span>")
+		return
+	return ..()
+
+/obj/vehicle/sealed/proc/remove_key(mob/user)
+	if(!inserted_key)
+		return
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: why is this return above the to_chat

File: code/modules/vehicles/entered.dm [hunk 12]
```diff
@@ -35,22 +35,74 @@
...
+		to_chat(user, "<span class='notice'>There is no key in \the [src]!</span>")
+	if(!is_occupant(user) || !(occupants[user] & VEHICLE_CONTROL_DRIVE))
+		to_chat(user, "<span class='notice'>You must be driving \the [src] to remove [src]'s key!</span>")
+		return
+	to_chat(user, "<span class='notice'>You remove \the [inserted_key] from \the [src].</span>")
+	inserted_key.forceMove(drop_location())
+	user.put_in_hands(inserted_key)
```
Comment (kevinz000) [inline code/modules/vehicles/entered.dm]: make sure this won't act weirdly with someone who's inside the car or somehow triggers this through TK. Once again, this is better unified with the existing proc for key insertion/removal.

File: code/modules/vehicles/car.dm [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+			DumpSpecificMobs(VEHICLE_CONTROL_KIDNAPPED)
+		else
+			to_chat(user, "<span class='notice'>It seems \the [src]'s trunk was empty.</span>")
+
+/obj/vehicle/sealed/car/proc/mob_try_forced_enter(mob/forcer, mob/M, silent = FALSE)
+	if(!istype(M))
+		return FALSE
+	if(occupant_amount() >= max_occupants)
+		return FALSE
+	if(do_after(forcer, get_enter_delay(M), target = src))
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: is there sanity checks for if the victim moves away?

File: code/modules/vehicles/car.dm [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+		return TRUE
+	mob_exit(M, silent)
+
+/obj/vehicle/sealed/car/after_move(direction)
+	if(world.time < last_enginesound_time + engine_sound_length)
+		return
+	last_enginesound_time = world.time
+	playsound(src, engine_sound, 100, TRUE)
+
+/obj/vehicle/sealed/car/attack_hand(mob/living/user)
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: needs to respect parent . = ..() return value for attack hand component signal.

File: code/modules/vehicles/car.dm [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+	if(!(traits & CAN_KIDNAP))
+		return
+	to_chat(user, "<span class='notice'>You start opening \the [src]'s trunk.</span>")
+	if(do_after(user, 30))
+		if(return_amount_of_controllers_with_flag(VEHICLE_CONTROL_KIDNAPPED))
+			to_chat(user, "<span class='notice'>The people stuck in \the [src]'s trunk all come tumbling out.</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: \the

File: code/modules/vehicles/car.dm [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+			DumpSpecificMobs(VEHICLE_CONTROL_KIDNAPPED)
+		else
+			to_chat(user, "<span class='notice'>It seems \the [src]'s trunk was empty.</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: \the

File: code/modules/vehicles/car.dm [hunk 17]
```diff
@@ -0,0 +1,122 @@
+#define CAN_KIDNAP 1
+
+/obj/vehicle/sealed/car
+	var/traits = NONE //Bitflag for special behavior such as kidnapping
+	var/engine_sound = 'sound/vehicles/carrev.ogg'
+	var/last_enginesound_time
+	var/engine_sound_length = 20 //Set this to the length of the engine sound
+	layer = ABOVE_MOB_LAYER
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: inherited variables before defined variables

File: code/modules/vehicles/car.dm [hunk 18]
```diff
@@ -0,0 +1,122 @@
+#define CAN_KIDNAP 1
```
[the remark this answers is not in the document]
Comment (kevinz000) [reply code/modules/vehicles/car.dm]: also shouldn't these be in flags.dm or something..?
also please make a varedit thing in glob.bitfields for admin use

File: code/modules/vehicles/car.dm [hunk 19]
```diff
@@ -0,0 +1,122 @@
+#define CAN_KIDNAP 1
+
+/obj/vehicle/sealed/car
+	var/traits = NONE //Bitflag for special behavior such as kidnapping
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: very generic name for a vehicle specific variable

File: code/modules/vehicles/car.dm [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+	if(traits & CAN_KIDNAP)
+		initialize_controller_action_type(/datum/action/vehicle/sealed/DumpKidnappedMobs, VEHICLE_CONTROL_DRIVE)
+
+/obj/vehicle/sealed/car/MouseDrop_T(atom/dropping, mob/M)
+	if(!usr.canmove || usr.stat || usr.restrained())
+		return FALSE
+	if(ismob(dropping) && M != dropping)
+		var/mob/D = dropping
+		M.visible_message("<span class='warning'>[M] starts forcing [D] into \the [src]!</span>")
+		mob_try_forced_enter(M, D)
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: i do not see a single Adjacent() or related check, hope you know what you're doing.

File: code/modules/vehicles/car.dm [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+	initialize_controller_action_type(/datum/action/vehicle/sealed/remove_key, VEHICLE_CONTROL_DRIVE)
+	if(traits & CAN_KIDNAP)
+		initialize_controller_action_type(/datum/action/vehicle/sealed/DumpKidnappedMobs, VEHICLE_CONTROL_DRIVE)
+
+/obj/vehicle/sealed/car/MouseDrop_T(atom/dropping, mob/M)
+	if(!usr.canmove || usr.stat || usr.restrained())
+		return FALSE
+	if(ismob(dropping) && M != dropping)
+		var/mob/D = dropping
+		M.visible_message("<span class='warning'>[M] starts forcing [D] into \the [src]!</span>")
```
Comment (kevinz000) [inline code/modules/vehicles/car.dm]: \the

File: code/modules/vehicles/car.dm [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+
+/obj/vehicle/sealed/car/clowncar
+	name = "clown car"
+	desc = "How someone could even fit in there is beyond me."
+	icon_state = "clowncar"
+	max_integrity = 500
+	armor = list("melee" = 70, "bullet" = 40, "laser" = 40, "energy" = 0, "bomb" = 30, "bio" = 0, "rad" = 0, "fire" = 80, "acid" = 80)
+	enter_delay = 20
+	max_occupants = 50
+	movedelay = 0.6
```
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: we'll find out

File: code/modules/vehicles/car.dm (hunk 18 above)
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: you put them there you numpty!

[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: you put flags in the wrong spot too but yeah i forgot to move these

File: code/modules/vehicles/car.dm [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+	if(ismob(dropping) && M != dropping)
+		var/mob/D = dropping
+		M.visible_message("<span class='warning'>[M] starts forcing [D] into \the [src]!</span>")
+		mob_try_forced_enter(M, D)
+	return ..()
+
+/obj/vehicle/sealed/car/mob_try_exit(mob/M, mob/user, silent = FALSE)
+	if(M == user && occupants[M] & VEHICLE_CONTROL_KIDNAPPED)
+		to_chat(user, "<span class='notice'>You push against the back of \the [src] trunk to try and get out.</span>")
+		if(!do_after(user, 200, target = src))
```
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: magic numbers

File: code/modules/vehicles/car.dm [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+	anchored = TRUE
+
+/obj/vehicle/sealed/car/generate_actions()
+	. = ..()
+	initialize_controller_action_type(/datum/action/vehicle/sealed/remove_key, VEHICLE_CONTROL_DRIVE)
+	if(traits & CAN_KIDNAP)
+		initialize_controller_action_type(/datum/action/vehicle/sealed/DumpKidnappedMobs, VEHICLE_CONTROL_DRIVE)
+
+/obj/vehicle/sealed/car/MouseDrop_T(atom/dropping, mob/M)
+	if(!usr.canmove || usr.stat || usr.restrained())
```
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: lol copy paste memes

File: code/modules/vehicles/car.dm [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+		return
+	to_chat(user, "<span class='notice'>You start opening \the [src]'s trunk.</span>")
+	if(do_after(user, 30))
+		if(return_amount_of_controllers_with_flag(VEHICLE_CONTROL_KIDNAPPED))
+			to_chat(user, "<span class='notice'>The people stuck in \the [src]'s trunk all come tumbling out.</span>")
+			DumpSpecificMobs(VEHICLE_CONTROL_KIDNAPPED)
+		else
+			to_chat(user, "<span class='notice'>It seems \the [src]'s trunk was empty.</span>")
+
+/obj/vehicle/sealed/car/proc/mob_try_forced_enter(mob/forcer, mob/M, silent = FALSE)
```
[the remark this answers is not in the document]
Comment (Qustinnus) [reply code/modules/vehicles/car.dm]: >:^)

**Comment by MrDoomBringer** [comment]
jesus kevin thats a lot of review

**Comment by Affurit** [comment]
Making an old man clown just to put people in my car.

**Pull request merged**